### PR TITLE
feat(plugin): GifCollections

### PR DESCRIPTION
**Pull request by Syncxv**
basically it lets you organize your gifs / images into collections

would love some feedback and improvements

https://user-images.githubusercontent.com/47534062/215351512-1ed9c95f-0579-41ce-8d4d-5bc3f101e0bc.mp4

todo:
-   [x] option to remove a collection
-   [x] option to remove a gif from collection
-   [x] use data store instead of localstorage
-   [x] figure out width and height of gifs :|
-   [x] support vidoes
-   [x] fix the flash that happens in the gif viewer thingy
-   [x] add context menu for messages with multiple vidoes
-   [x] option to export your collections
-   [x] option to import your collections
-   [x] add collection context menu in the gif picker

### CODE REVIEW

**Comment by Vendicated** [comment]
hi, is this done yet?

**Comment by Syncxv** [comment]
yeah its finished :D

i forgort to say it finished was whoops

**Comment by Vendicated** [comment]
okay cool, wasn't so sure because your last commits all sound like this is a WIP

**Comment by Vendicated** [comment]
this is a really cool plugin btw, but now write an AI that will sort my ~200 favourited gifs into collections

**Comment by Syncxv** [comment]
ayo

File: src/plugins/gifCollections/CollectionManager.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    return collections.find(c => c.name === name);
+};
+
+export const getCachedCollection = (name: string) => cache_collections.find(c => c.name === name);
+
+export const createCollection = async (name: string, gifs: Gif[]) => {
+
+    const collections = await DataStore.get<Collection[]>(DATA_COLLECTION_NAME) ?? [];
+    const duplicateCollection = collections.find(c => c.name === `gc:${name}`);
+    // TODO: notify user instead of just console.warn
```
Comment (Vendicated) [inline src/plugins/gifCollections/CollectionManager.ts:41]: I would just show a Toast for this, you can find many examples of toasts in the codebase

Comment (Syncxv) [reply src/plugins/gifCollections/CollectionManager.ts:41]: crap i forgot about that 1 sec

File: src/plugins/gifCollections/CollectionManager.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if (duplicateCollection) return console.warn("collection already exists");
+
+    collections.push({
+        name: `gc:${name}`,
+        src: gifs.length ? gifs[gifs.length - 1].src : "",
+        format: getFormat(gifs[gifs.length - 1].src),
```
Comment (Vendicated) [inline src/plugins/gifCollections/CollectionManager.ts:47]: what's the point of checking length one line above if you're not gonna check here?

File: src/plugins/gifCollections/CollectionManager.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    collections[collectionIndex].format = getFormat(gif.src);
+
+    await DataStore.set(DATA_COLLECTION_NAME, collections);
+    return await refreshCacheCollection();
+
+};
+
+export const removeFromCollection = async (id: string) => {
+    const collections = await DataStore.get<Collection[]>(DATA_COLLECTION_NAME) ?? [];
+    const collectionIndex = collections.findIndex(c => c.gifs.findIndex(g => g.id === id) !== -1);
```
Comment (Vendicated) [inline src/plugins/gifCollections/CollectionManager.ts:72]:
```suggestion
    const collectionIndex = collections.findIndex(c => c.gifs.some(g => g.id === id));
```

File: src/plugins/gifCollections/index.tsx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { Devs } from "@utils/constants";
+import { makeLazy } from "@utils/misc";
+import { ModalContent, ModalFooter, ModalHeader, ModalRoot, openModal } from "@utils/modal";
+import definePlugin from "@utils/types";
+import { Alerts, Button, ContextMenu, FluxDispatcher, Forms, Menu, React, TextInput } from "@webpack/common";
+import { Settings } from "Vencord";
```
Comment (Vendicated) [inline src/plugins/gifCollections/index.tsx:24]: bad
```suggestion
import { Settings } from "@api/Settings";
```

File: src/plugins/gifCollections/utils/uuidv4.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+export function uuidv4() {
+    let d = new Date().getTime();
+    if (typeof performance !== "undefined" && typeof performance.now === "function") {
```
Comment (Vendicated) [inline src/plugins/gifCollections/utils/uuidv4.ts:21]:
```suggestion
```

File: src/plugins/gifCollections/utils/getUrlExtension.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+export function getUrlExtension(url: string) {
+    return url.split(/[#?]/)[0].split(".").pop()?.trim();
```
Comment (Vendicated) [inline src/plugins/gifCollections/utils/getUrlExtension.ts:20]:
```suggestion
    return new URL(url).pathname.split(".").pop();
```

File: src/plugins/gifCollections/utils/getFormat.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { Format } from "../types";
+import { getUrlExtension } from "./getUrlExtension";
+
+export function getFormat(url: string) {
+    const exstension = getUrlExtension(url);
```
Comment (Vendicated) [inline src/plugins/gifCollections/utils/getFormat.ts:23]: 🤨

File: src/plugins/gifCollections/index.tsx [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+        const gif: Gif = {
+            id: uuidv4(),
+            height,
+            width,
+            src,
+            url
+        };
+
+        console.log(gif);
```
Comment (Vendicated) [inline src/plugins/gifCollections/index.tsx:174]:
```suggestion
```

File: src/plugins/gifCollections/index.tsx [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+                    }}
+                />
+            </Menu.MenuItem>
+        );
+    }
+});
+
+// stolen from spotify controls
+const RemoveItemContextMenu = ({ type, nameOrId, onConfirm }: { type: "gif" | "collection", nameOrId: string, onConfirm: () => void; }) => (
+    <Menu.ContextMenu
```
Comment (Vendicated) [inline src/plugins/gifCollections/index.tsx:216]: you need to depend on the MenuItemDeobfuscatorApi plugin to use context menus

**Review by Vendicated — CHANGES_REQUESTED** [review]

**Comment by Syncxv** [comment]
ok there seems to be an issue with the menu. target is undefined so i cant get the height and widths of the gif. ima try fix it gimme a sec

File: src/plugins/gifCollections/utils/getFormat.ts (hunk 7 above)
Comment (Syncxv) [reply src/plugins/gifCollections/utils/getFormat.ts:23]: i dont get it 
what should i do?

File: src/plugins/gifCollections/CollectionManager.ts (hunk 2 above)
Comment (Syncxv) [reply src/plugins/gifCollections/CollectionManager.ts]: is it good now?

can i resolve

File: src/plugins/gifCollections/utils/getFormat.ts (hunk 7 above)
Comment (Vendicated) [reply src/plugins/gifCollections/utils/getFormat.ts:23]: typo silly

**Comment by Syncxv** [comment]
Ready

**Comment by somerandomcloud** [comment]
this pr is BAssed as fucK
4/5

**Pull request closed without merging**